File: src/main/java/edu/cornell/cs/apl/viaduct/security/Label.java
package edu.cornell.cs.apl.viaduct.security;

import com.google.auto.value.AutoValue;
import edu.cornell.cs.apl.viaduct.algebra.FreeDistributiveLattice;
import edu.cornell.cs.apl.viaduct.algebra.Lattice;

/**
 * A lattice for information flow security. This is a standard bounded lattice that additionally
 * supports confidentiality and integrity projections. Information flows from less restrictive
 * contexts to more restrictive ones.
 *
 * <p>{@link #top()}, {@link #bottom()}, {@link #meet(Label)}, and {@link #join(Label)} talk about
 * information flow.
 *
 * <p>{@link #weakest()}, {@link #strongest()}, {@link #and(Label)}, and {@link #or(Label)} talk
 * about trust.
 */
@AutoValue
public abstract class Label implements Lattice<Label>, TrustLattice<Label> {
  // public, untrusted
  private static final Label WEAKEST =
      create(FreeDistributiveLattice.top(), FreeDistributiveLattice.top());

  // secret, trusted
  private static final Label STRONGEST =
      create(FreeDistributiveLattice.bottom(), FreeDistributiveLattice.bottom());

  // public, trusted
  private static final Label BOTTOM =
      create(weakest().getConfidentialityComponent(), strongest().getIntegrityComponent());

  // secret, untrusted
  private static final Label TOP =
      create(strongest().getConfidentialityComponent(), weakest().getIntegrityComponent());

  /** Constructs a label corresponding to a single given principal. */
  public static Label create(Principal principal) {
    final FreeDistributiveLattice<Principal> component = FreeDistributiveLattice.create(principal);
    return create(component, component);
  }

  /** Constructs a label given the confidentiality and integrity components. */
  public static Label create(
      FreeDistributiveLattice<Principal> confidentiality,
      FreeDistributiveLattice<Principal> integrity) {
    return new AutoValue_Label(confidentiality, integrity);
  }

  /** Constructs a label given only the confidentiality component. Integrity is set to minimum. */
  public static Label fromConfidentiality(FreeDistributiveLattice<Principal> confidentiality) {
    return Label.create(confidentiality, Label.weakest().getIntegrityComponent());
  }

  /** Construct a label given only the integrity component. Confidentiality is set to minimum. */
  public static Label fromIntegrity(FreeDistributiveLattice<Principal> integrity) {
    return Label.create(Label.weakest().getConfidentialityComponent(), integrity);
  }

  /** The least restrictive data policy, i.e. public and trusted. */
  public static Label bottom() {
    return BOTTOM;
  }

  /** The most restrictive data policy, i.e. secret and untrusted. */
  public static Label top() {
    return TOP;
  }

  /**
   * The least powerful principal, i.e. public and untrusted.
   *
   * <p>This is represented as 1, and is the unit for {@link #and(Label)}.
   */
  public static Label weakest() {
    return WEAKEST;
  }

  /**
   * The most powerful principal, i.e. secret and trusted.
   *
   * <p>This is represented as 0, and is the unit for {@link #or(Label)}.
   */
  public static Label strongest() {
    return STRONGEST;
  }

  /**
   * Return the confidentiality component in the underlying lattice.
   *
   * <p>Unlike {@link #confidentiality()}, the result is not a {@link Label}.
   */
  public abstract FreeDistributiveLattice<Principal> getConfidentialityComponent();

  /**
   * Return the integrity component in the underlying lattice.
   *
   * <p>Unlike {@link #integrity()}, the result is not a {@link Label}.
   */
  public abstract FreeDistributiveLattice<Principal> getIntegrityComponent();

  /** Check if information flow from {@code this} to {@code that} is safe. */
  public final boolean flowsTo(Label that) {
    return that.confidentiality()
        .and(this.integrity())
        .actsFor(this.confidentiality().and(that.integrity()));
  }

  @Override
  public final boolean lessThanOrEqualTo(Label that) {
    return this.flowsTo(that);
  }

  @Override
  public final Label join(Label with) {
    return this.and(with).confidentiality().and(this.or(with).integrity());
  }

  @Override
  public final Label meet(Label with) {
    return this.or(with).confidentiality().and(this.and(with).integrity());
  }

  /**
   * The confidentiality component.
   *
   * <p>Keeps confidentiality the same while setting integrity to the weakest level.
   */
  public final Label confidentiality() {
    return create(this.getConfidentialityComponent(), weakest().getIntegrityComponent());
  }

  /**
   * The integrity component.
   *
   * <p>Keeps integrity the same while setting confidentiality to the weakest level.
   */
  public final Label integrity() {
    return create(weakest().getConfidentialityComponent(), this.getIntegrityComponent());
  }

  @Override
  public final boolean actsFor(Label other) {
    return this.getConfidentialityComponent().lessThanOrEqualTo(other.getConfidentialityComponent())
        && this.getIntegrityComponent().lessThanOrEqualTo(other.getIntegrityComponent());
  }

  @Override
  public final Label and(Label with) {
    final FreeDistributiveLattice<Principal> confidentiality =
        this.getConfidentialityComponent().meet(with.getConfidentialityComponent());
    final FreeDistributiveLattice<Principal> integrity =
        this.getIntegrityComponent().meet(with.getIntegrityComponent());
    return create(confidentiality, integrity);
  }

  @Override
  public final Label or(Label with) {
    final FreeDistributiveLattice<Principal> confidentiality =
        this.getConfidentialityComponent().join(with.getConfidentialityComponent());
    final FreeDistributiveLattice<Principal> integrity =
        this.getIntegrityComponent().join(with.getIntegrityComponent());
    return create(confidentiality, integrity);
  }

  /**
   * Switch the confidentiality and integrity components.
   *
   * <p>This is used to enforce robust declassification and transparent endorsement (a.k.a.
   * non-malleable information flow).
   */
  public final Label swap() {
    return Label.create(getIntegrityComponent(), getConfidentialityComponent());
  }

  @Override
  public final String toString() {
    final String confidentialityStr = this.getConfidentialityComponent().toString();
    final String integrityStr = this.getIntegrityComponent().toString();

    String expression;
    if (this.equals(weakest())) {
      expression = "";

    } else if (this.getConfidentialityComponent().equals(this.getIntegrityComponent())) {
      expression = confidentialityStr;

    } else if (this.equals(this.confidentiality())) {
      expression = confidentialityStr + "->";

    } else if (this.equals(this.integrity())) {
      expression = integrityStr + "<-";

    } else {
      expression = String.format("%s-> ∧ %s<-", confidentialityStr, integrityStr);
    }

    return String.format("{%s}", expression);
  }
}

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/security/Label.kt
package edu.cornell.cs.apl.viaduct.security

import edu.cornell.cs.apl.prettyprinting.Document
import edu.cornell.cs.apl.prettyprinting.PrettyPrintable
import edu.cornell.cs.apl.prettyprinting.Style
import edu.cornell.cs.apl.viaduct.algebra.FreeDistributiveLattice
import edu.cornell.cs.apl.viaduct.algebra.Lattice
import edu.cornell.cs.apl.viaduct.security.Label.Companion.bottom
import edu.cornell.cs.apl.viaduct.security.Label.Companion.strongest
import edu.cornell.cs.apl.viaduct.security.Label.Companion.top
import edu.cornell.cs.apl.viaduct.security.Label.Companion.weakest

/**
 * A lattice for information flow security. This is a standard bounded lattice that additionally
 * supports confidentiality and integrity projections. Information flows from less restrictive
 * contexts to more restrictive ones.
 *
 * [top], [bottom], [meet], and [join] talk about information flow.
 *
 * [weakest], [strongest], [and], and [or] talk about trust.
 */
data class Label(
    /**
     * The confidentiality component in the underlying lattice.
     *
     * Unlike [confidentiality], the result is not a [Label].
     */
    val confidentialityComponent: FreeDistributiveLattice<Principal>,

    /**
     * The integrity component in the underlying lattice.
     *
     * Unlike [integrity], the result is not a [Label].
     */
    val integrityComponent: FreeDistributiveLattice<Principal>
) : Lattice<Label>, TrustLattice<Label>, PrettyPrintable {
    /**
     * The confidentiality component.
     *
     * Keeps confidentiality the same while setting integrity to the weakest level.
     */
    fun confidentiality(): Label =
        fromConfidentiality(confidentialityComponent)

    /**
     * The integrity component.
     *
     * Keeps integrity the same while setting confidentiality to the weakest level.
     */
    fun integrity(): Label = fromIntegrity(integrityComponent)

    /** Check if information flow from `this` to [that] is safe. */
    infix fun flowsTo(that: Label): Boolean =
        (that.confidentiality() and this.integrity()) actsFor (this.confidentiality() and that.integrity())

    override fun lessThanOrEqualTo(that: Label): Boolean =
        this.flowsTo(that)

    override infix fun join(that: Label): Label =
        (this.confidentiality() and that.confidentiality()) and (this.integrity() or that.integrity())

    override infix fun meet(that: Label): Label =
        (this.confidentiality() or that.confidentiality()) and (this.integrity() and that.integrity())

    override infix fun actsFor(that: Label): Boolean =
        this.confidentialityComponent.lessThanOrEqualTo(that.confidentialityComponent) &&
            this.integrityComponent.lessThanOrEqualTo(that.integrityComponent)

    override infix fun and(that: Label): Label {
        val confidentiality = this.confidentialityComponent.meet(that.confidentialityComponent)
        val integrity = this.integrityComponent.meet(that.integrityComponent)
        return Label(confidentiality, integrity)
    }

    override infix fun or(that: Label): Label {
        val confidentiality = this.confidentialityComponent.join(that.confidentialityComponent)
        val integrity = this.integrityComponent.join(that.integrityComponent)
        return Label(confidentiality, integrity)
    }

    /**
     * Switch the confidentiality and integrity components.
     *
     * This is used to enforce robust declassification and transparent endorsement
     * (a.k.a. non-malleable information flow).
     */
    fun swap(): Label =
        Label(integrityComponent, confidentialityComponent)

    override fun toString(): String {
        val confidentialityStr = confidentialityComponent.toString()
        val integrityStr = integrityComponent.toString()
        val expression =
            when {
                confidentialityComponent == integrityComponent ->
                    confidentialityStr
                this == this.confidentiality() ->
                    "$confidentialityStr->"
                this == this.integrity() ->
                    "$integrityStr<-"
                else ->
                    "$confidentialityStr-> ∧ $integrityStr<-"
            }
        return "{$expression}"
    }

    // TODO: make asDocument primitive and remove toString
    override val asDocument: Document
        get() = Document(this.toString())

    companion object {
        /**
         * The least powerful principal, i.e. public and untrusted.
         *
         * This is the unit for [and].
         */
        @JvmStatic
        val weakest: Label = Label(FreeDistributiveLattice.top(), FreeDistributiveLattice.top())

        /**
         * The most powerful principal, i.e. secret and trusted.
         *
         * This is the unit for [or].
         */
        @JvmStatic
        val strongest: Label = Label(FreeDistributiveLattice.bottom(), FreeDistributiveLattice.bottom())

        /**
         * The least restrictive data policy, i.e. public and trusted.
         *
         * This is the unit for [join].
         */
        @JvmStatic
        val bottom: Label = Label(weakest.confidentialityComponent, strongest.integrityComponent)

        /**
         * The most restrictive data policy, i.e. secret and untrusted.
         *
         * This is the unit for [meet].
         */
        @JvmStatic
        val top: Label = Label(strongest.confidentialityComponent, weakest.integrityComponent)

        /** Returns the label representing the authority of the given principal. */
        @JvmStatic
        operator fun invoke(principal: Principal): Label {
            val component = FreeDistributiveLattice(principal)
            return Label(component, component)
        }

        /** Constructs a label given only the confidentiality component. Integrity is set to minimum. */
        @JvmStatic
        fun fromConfidentiality(confidentiality: FreeDistributiveLattice<Principal>): Label =
            Label(confidentiality, weakest.integrityComponent)

        /** Constructs a label given only the integrity component. Confidentiality is set to minimum. */
        @JvmStatic
        fun fromIntegrity(integrity: FreeDistributiveLattice<Principal>): Label =
            Label(weakest.confidentialityComponent, integrity)
    }
}

/** The display style of [Label] specific operators such as [Label.confidentiality]. */
object LabelOperatorStyle : Style
